src/node/build.ts

```diff
@@ -124,38 +122,79 @@ export async function build(cliOptions: Partial<ViteSSGOptions> = {}, viteConfig
   // uniq
   routesPaths = Array.from(new Set(routesPaths))
 
-  buildLog('Rendering Pages...', routesPaths.length)
+  buildLog('Rendering SSR app for each route...', routesPaths.length)
 
   const critters = crittersOptions !== false ? await getCritters(outDir, crittersOptions) : undefined
   if (critters)
     console.log(`${gray('[vite-ssg]')} ${blue('Critical CSS generation enabled via `critters`')}`)
 
-  const ssrManifest: Manifest = JSON.parse(await fs.readFile(join(out, 'ssr-manifest.json'), 'utf-8'))
-  let indexHTML = await fs.readFile(join(out, 'index.html'), 'utf-8')
-  indexHTML = rewriteScripts(indexHTML, script)
-
   const { renderToString }: typeof import('vue/server-renderer') = await import('vue/server-renderer')
 
-  // @ts-expect-error just ignore it hasn't exports on its package
-  // eslint-disable-next-line new-cap
-  const queue = new PQueue.default({ concurrency })
+  const appRenderQueue = new PQueue({ concurrency })
+
+  // First we want to render the SSR app for all routes.
+  const appRenderResults : {filename: string, route: string, ctx: SSRContext, appCtx: ViteSSGContext<true> }[] = [];
 
   for (const route of routesPaths) {
-    queue.add(async () => {
+    appRenderQueue.add(async () => {
       try {
         const appCtx = await createApp(false, route) as ViteSSGContext<true>
-        const { app, router, head, initialState, triggerOnSSRAppRendered, transformState = serializeState } = appCtx
+        const { app, router, triggerOnSSRAppRendered } = appCtx
 
         if (router) {
           await router.push(route)
           await router.isReady()
         }
 
-        const transformedIndexHTML = (await onBeforePageRender?.(route, indexHTML, appCtx)) || indexHTML
-
         const ctx: SSRContext = {}
         const appHTML = await renderToString(app, ctx)
         await triggerOnSSRAppRendered?.(route, appHTML, appCtx)
+
+        const filename = getRouteFileName(route, dirStyle, ssgOut)
+
+        // Write file to keep memory usage down
+        await fs.ensureDir(dirname(filename))
+        await fs.writeFile(filename, appHTML, 'utf-8')
+
+        appRenderResults.push({ filename, route, ctx, appCtx })
+      }
+      catch (err: any) {
+        throw new Error(`${gray('[vite-ssg]')} ${red(`Error rendering SSR app: ${cyan(route)}`)}\n${err.stack}`)
+      }
+    })
+  }
+
+  await appRenderQueue.start().onIdle()
+
+  // Now that the SSR apps have run (which may have created client assets), we can run the client build. 
+  buildLog('Build for client...')
+  await viteBuild(mergeConfig(viteConfig, {
```

Comment:
should we reset `process.env.VITE_SSG` and  `process.env.SSR` before running the client build?

Reply:
how about deregistering `jsdom`?

Reply:
Is it really necessary to reverse the build order?

Reply:
If we are to support use cases such as the one I have provided, then yes, reversing the build order is required.

Reply:
> how about deregistering `jsdom`?
> 
> EDIT: check this line https://github.com/antfu/vite-ssg/blob/main/src/client/index.ts#L24



> how about deregistering `jsdom`?
> 
> EDIT: check this line https://github.com/antfu/vite-ssg/blob/main/src/client/index.ts#L24

I'm not sure I understand the feedback. That line looks like it's checking for the existence of the window object in order to determine whether it should do a client build. Nothing about that would be changed by my PR, as far as I can tell.